pkg/controller/servicemesh/controlplane/controller_test.go

```diff
@@ -1,104 +1,200 @@
-	instanceReconciler := reconciler.getOrCreateReconciler(&v1.ServiceMeshControlPlane{})
+func TestUpdateReadinessInvokedWhenInstanceFullyReconciled(t *testing.T) {
+	controlPlane := newControlPlane()
+	controlPlane.Status.ReconciledVersion = maistrav1.CurrentReconciledVersion(controlPlane.Generation)
+	controlPlane.Status.Conditions = append(controlPlane.Status.Conditions, maistrav1.Condition{
+		Type:               maistrav1.ConditionTypeReconciled,
+		Status:             maistrav1.ConditionStatusTrue,
+		Reason:             "",
+		Message:            "",
+		LastTransitionTime: oneMinuteAgo,
+	})
+
+	_, _, _, r := createClientAndReconciler(t, controlPlane)
+	assertReconcileSucceeds(r, t)
+
+	assert.True(instanceReconciler.updateReadinessInvoked, "Expected UpdateReadiness() to be invoked on instance reconciler", t)
+	assert.False(instanceReconciler.reconcileInvoked, "Expected Reconcile() to NOT be invoked on instance reconciler", t)
+}
 
-	_, err := instanceReconciler.recursivelyApplyTemplates(v1.ControlPlaneSpec{Template: "visited"}, "", sets.NewString("visited"))
+func TestReconcileDoesNothingWhenResourceIsNotFound(t *testing.T) {
+	_, tracker, _, r := createClientAndReconciler(t)
+	assertReconcileSucceeds(r, t)
+	test.AssertNumberOfWriteActions(t, tracker.Actions(), 0)
+}
+
+func TestReconcileFailsWhenGetResourceFails(t *testing.T) {
+	_, tracker, _, r := createClientAndReconciler(t)
+	tracker.AddReactor(test.ClientFailsOn("get", "servicemeshcontrolplanes"))
+
+	assertReconcileFails(r, t)
+
+	test.AssertNumberOfWriteActions(t, tracker.Actions(), 0)
+}
+
+func createClientAndReconciler(t *testing.T, clientObjects ...runtime.Object) (client.Client, *test.EnhancedTracker, record.EventRecorder, *ControlPlaneReconciler) {
+	cl, enhancedTracker := test.CreateClient(clientObjects...)
+	fakeEventRecorder := &record.FakeRecorder{}
+
+	r := newReconciler(cl, scheme.Scheme, fakeEventRecorder, "istio-operator")
+	r.instanceReconcilerFactory = NewFakeInstanceReconciler
+	instanceReconciler = &fakeInstanceReconciler{}
+	return cl, enhancedTracker, fakeEventRecorder, r
+}
+
+type fakeInstanceReconciler struct {
+	reconcileInvoked       bool
+	updateReadinessInvoked bool
+	deleteInvoked          bool
+	finished               bool
+}
+
+func NewFakeInstanceReconciler(controllerResources common.ControllerResources, instance *maistrav1.ServiceMeshControlPlane) ControlPlaneInstanceReconciler {
+	return instanceReconciler
+}
+
+func (r *fakeInstanceReconciler) Reconcile() (reconcile.Result, error) {
+	r.reconcileInvoked = true
+	return reconcile.Result{}, nil
+}
+
+func (r *fakeInstanceReconciler) UpdateReadiness() error {
+	r.updateReadinessInvoked = true
+	return nil
+}
+
+func (r *fakeInstanceReconciler) Delete() error {
+	r.deleteInvoked = true
+	return nil
+}
+
+func (r *fakeInstanceReconciler) SetInstance(instance *maistrav1.ServiceMeshControlPlane) {
+}
+
+func (r *fakeInstanceReconciler) IsFinished() bool {
+	return r.finished
```

Comment:
Do we ever set this to `true` in the fake reconciler? It's not super important here, but the `defer` calls that are supposed to clean up probably don't do much

Reply:
Good catch. I added the {{finished}} field because I planned to also test {{getOrCreateReconciler()}} and {{deleteReconcilerIfFinished()}}. Then I apparently forgot to add those tests. I'll merge this PR now and add those tests when I get back on Friday.

Reply:
Sounds good. Enjoy your time off!